### PR TITLE
Add support for assignment of component or element instance to vars and support for #name var shorthand.

### PR DESCRIPTION
**Pull request by jelbourn**
Currently missing unit tests for View changes, as I'm not sure about the best way to test them in isolation.
Due to unfamiliarity with the codebase and the framework overall, I may have no considered some test cases.

### CODE REVIEW

**Comment by mhevery** [comment]
Squash to single SHA, rebase to master, and push.

File: modules/angular2/src/core/compiler/element_injector.js [hunk 1]
```diff
@@ -211,10 +211,22 @@ export class ProtoElementInjector  {
   index:int;
   view:View;
   distanceToParent:number;
+
+  /** Whether the element is exported as $implicit. */
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/element_injector.js]: maybe I am missing something, but why can't this data live on the element binder, just like events? I am not convinced creating Element Injectors is necessary just to support var-.

https://github.com/angular/angular/blob/master/modules/angular2/src/core/compiler/element_binder.js#L22

File: modules/angular2/src/core/compiler/view.js [hunk 2]
```diff
@@ -151,6 +151,16 @@ export class View {
...
         elementInjector.instantiateDirectives(appInjector, shadowDomAppInjector, this.preBuiltObjects[i]);
+
+        // The exporting of $implicit is a special case. Since multiple elements will all export
+        // the different values as $implicit, directly assign $implicit bindings to the variable
+        // name.
+        var exportImplicitName = elementInjector.getExportImplicitName();
+        if (elementInjector.isExportingComponent()) {
+          this.context.set(exportImplicitName, elementInjector.getComponent());
+        } else if (elementInjector.isExportingElement()) {
+          this.context.set(exportImplicitName, elementInjector.getNgElement().domElement);
```
Comment (rkirov) [inline modules/angular2/src/core/compiler/view.js]: `this.preBuildObjects` is indexed the same as `elementInjectors`, so `this.preBuildObjects[i].element` should do the trick without adding extra API to `elementInjector`

Comment (mhevery) [inline modules/angular2/src/core/compiler/view.js]: If that is the case than we should do this.

Comment (jelbourn) [inline modules/angular2/src/core/compiler/view.js]: I'll make this change in the next PR.

**Pull request closed without merging**